### PR TITLE
Fixes to `write_h5ad_categorical()`

### LINKED ISSUE #154: Write HDF5 categoricals correctly using attributes
Currently, the information about whether a categorical is ordered is written as a dataset but it should be an attribute.

Should be a relatively easy fix but is complicated by writing of logical attributes not working in **{rhdf5}** (see https://github.com/grimbough/rhdf5/issues/136).

Other things to check:

- `categories` are written correctly as strings
- `codes` are written correctly
  - 0-indexed
  - `NA`s are converted to `-1`

### PR DESCRIPTION
**Pull request by lazappi**
- Write `ordered` as an attribute rather than a dataset
- Write `categories` as a string
- Write `codes` as 0-indexed integer to with `NA`s set to `-1`

Fixes #154

### CODE REVIEW

File: R/write_h5ad_helpers.R [hunk 1]
```diff
@@ -276,11 +276,25 @@ write_h5ad_string_array <- function(value, file, name, compression, version = "0
...
   write_h5ad_encoding(file, name, "categorical", version)
+
+  # Write ordered attribute
+  h5file <- rhdf5::H5Fopen(file)
+  on.exit(rhdf5::H5Fclose(h5file))
+
+  h5obj <- rhdf5::H5Gopen(h5file, name)
+  on.exit(rhdf5::H5Gclose(h5obj), add = TRUE)
+
+  rhdf5::h5writeAttribute(is.ordered(value), h5obj, "ordered", asScalar = TRUE)
```
[the remark this answers is not in the document]
Comment (LouiseDck) [reply R/write_h5ad_helpers.R:299]: That should be possible, no?

**Comment by rcannood** [comment]
Or, I can merge #138 into main first -- since it doesn't seem to be failing at this moment in time. Then we can merge main back into this PR.

File: R/write_h5ad_helpers.R (hunk 1 above)
[the remark this answers is not in the document]
Comment (lazappi) [reply R/write_h5ad_helpers.R:299]: Yeah, I should be able to do this. This pattern is used enough times now to justify a function.

**Comment by rcannood** [comment]
This PR breaks the following test:

    ── Failure ('test-HDF5-write.R:91:3'): Writing H5AD categoricals works ─────────
    hdf5_path_exists(h5ad_file, "/categorical/ordered") is not TRUE

This makes sense, since `ordered` is no longer a path but an attribute.

**Comment by lazappi** [comment]
I think the remaining error here is because of the **{rhdf5}** problem. I'm not really sure what to do about that.

**Comment by lazappi** [comment]
> So maybe one question is: if something like `write_h5ad_categorical` fails, should we:
> 
> * throw an error
> * generate a warning but at least remove the `categorical` field
> * generate a warning but the data might be incomplete

I'm pretty sure any categorical data will show this error. We can't have an incomplete field because then Python **anndata** won't be able to read the file. The only solution I can think of is to try and fix this in **{rhdf5}** and/or add a function for writing logical attributes here. Temporarily maybe the best we can do is add a warning to `write_h5ad_categorical()` saying this is not yet implemented and do nothing. It's a fairly essential part of functionally though.

**Comment by lazappi** [comment]
> However, in any situation where a write fails, it would probably be best to remove the field and show a warning. WDYT? (No need to already implement this, let's discuss first)

I think that's reasonable in general (it's what **{zellkonverter}** does). Maybe we want an argument to control whether people get an error or a warning (for cases when you need to be sure things are written)?

> Yeah. I tried to have a look at `rhdf5`. Looking at the implementations in `h5py` and `hdf5r` , it seems we have to create an 8-bit enum with `{0: FALSE, 1: TRUE}`, but I'm not familiar enough with the `rhdf5` codebase to make these changes. I tried, but I didn't get to something that worked :P

Yeah, I think it's beyond me as well. I guess we should try to push for that issue to be resolved?

**Comment by lazappi** [comment]
@rcannood I think this should work now with the current devel version of **{rhdf5}** if you want to take a look at it. I suspect the CI will fail though.

**Review by rcannood — APPROVED** [review]

**Comment by lazappi** [comment]
Thanks for merging! We might want to bump the **{rhdf5}** version in `DESCRIPTION` to make it clearer what is required. I think I didn't do that already to avoid people needing to install Bioconductor devel.

**Comment by rcannood** [comment]
Good idea, thanks!

**Pull request merged**